### PR TITLE
mon/OSDMonitor: issue pool application related warning

### PR DESCRIPTION
**Pull request by xiexingguo**
https://github.com/ceph/ceph/pull/15763 says:
` In-use pools that are not associated to an application will generate a health warning.`

Ceph will unable to do this until this patch is applied.

### CODE REVIEW

File: src/mon/OSDMonitor.cc [hunk 1]
```diff
@@ -6429,6 +6429,12 @@ int OSDMonitor::prepare_command_pool_application(const string &prefix,
       return -EPERM;
     }
 
+    if (!app_exists) {
+      ss << "application '" << app << "' is not enabled on pool '" << pool_name
+         << "'";
+      return -ENOENT;
```
Comment (liewegas) [inline src/mon/OSDMonitor.cc]: we should return success in this case so that the command is idempotent

File: src/mon/OSDMonitor.cc [hunk 2]
```diff
@@ -6492,8 +6492,14 @@ int OSDMonitor::prepare_command_pool_application(const string &prefix,
 
     string key;
     cmd_getval(g_ceph_context, cmdmap, "key", key);
+    auto it = p.application_metadata[app].find(key);
+    if (it == p.application_metadata[app].end()) {
+      ss << "application '" << app << "' on pool '" << pool_name
+         << "' does not have key '" << key << "'";
+      return -ENOENT;
```
Comment (liewegas) [inline src/mon/OSDMonitor.cc]: return success here too

**Comment by liewegas** [comment]
http://pulpito.ceph.com/sage-2017-07-24_18:23:25-rados-wip-sage-testing-distro-basic-smithi/

we need to add whitelists to a bunch of tests

**Comment by dillaman** [comment]
There is existing application metadata health check code here [1]. If it's not working after the rebase to the refactored health warnings, it should be fixed and/or removed.

[1] https://github.com/ceph/ceph/blob/master/src/mon/OSDMonitor.cc#L3942

**Comment by liewegas** [comment]
That is actually dead code. It's generating an old-style health warning, but the block is guarded by >= luminous.  Let's remove it!

**Comment by dillaman** [comment]
@liewegas So it's dead code starting from https://github.com/ceph/ceph/blob/master/src/mon/OSDMonitor.cc#L3636 as well? I guess that would explain it.

**Comment by liewegas** [comment]
eek.. yup

**Comment by liewegas** [comment]
see also https://github.com/ceph/ceph/pull/16568

**Comment by liewegas** [comment]
added a commit to #16568 that removes the dead code

**Comment by liewegas** [comment]
retest this please

**Comment by liewegas** [comment]
retest this please

**Comment by xiexingguo** [comment]
retest this please

**Comment by liewegas** [comment]
retest this please

**Comment by dzafman** [comment]
@liewegas I couldn't reproduce this on my build machine, but I guess we need the same change for the run_mon in test_pidfile.

See #16635

**Comment by liewegas** [comment]
I dont' think this should affect the pidfile test; the locking error comes up before the mon is even started.

**Comment by liewegas** [comment]
http://pulpito.ceph.com/sage-2017-07-28_04:13:20-rados-wip-sage-testing-distro-basic-smithi/

**Pull request merged**